pkg/reader/usdc_reader.go

```diff
@@ -338,6 +360,148 @@ func (u evmUSDCMessageReader) recreateMessageTransmitterEvents(
 	return messageTransmitterEvents, nil
 }
 
+type solanaUSDCMessageReader struct {
+	lggr           logger.Logger
+	contractReader contractreader.Extended
+	cctpDestDomain map[uint64]uint32
+}
+
+// getMessageTokenData extracts token data from the CCTP MessageSent event.
+func (u solanaUSDCMessageReader) getMessageTokenData(
+	tokens map[MessageTokenID]cciptypes.RampTokenAmount,
+) (map[MessageTokenID]SourceTokenDataPayload, error) {
+	messageTransmitterEvents := make(map[MessageTokenID]SourceTokenDataPayload)
+
+	for id, token := range tokens {
+		sourceTokenPayload, err := extractABIPayload(token.ExtraData)
+		if err != nil {
+			return nil, err
+		}
+		messageTransmitterEvents[id] = *sourceTokenPayload
+	}
+	return messageTransmitterEvents, nil
+}
+
+// SolanaCCTPUSDCMessageEvent matches the onChain CcipCctpMessageSentEvent.
+type SolanaCCTPUSDCMessageEvent struct {
+	Discriminator       [8]byte
+	OriginalSender      [32]byte
+	RemoteChainSelector uint64
+	MsgTotalNonce       uint64
+	EventAddress        [32]byte
+	SourceDomain        uint32
+	CctpNonce           uint64
+	MessageSentBytes    []byte
+}
+
+// MessagesByTokenID retrieves the CCTP MessageSent events.
+func (u solanaUSDCMessageReader) MessagesByTokenID(
+	ctx context.Context,
+	source, dest cciptypes.ChainSelector,
+	tokens map[MessageTokenID]cciptypes.RampTokenAmount,
+) (map[MessageTokenID]cciptypes.Bytes, error) {
+	if len(tokens) == 0 {
+		return map[MessageTokenID]cciptypes.Bytes{}, nil
+	}
+	u.lggr.Debugw("Searching for Solana CCTP USDC logs",
+		"numExpected", len(tokens))
+
+	// Parse the extra data field to get the CCTP nonces and source domains.
+	cctpData, err := u.getMessageTokenData(tokens)
+	if err != nil {
+		return nil, err
+	}
+
+	// Query the token pool contract for the MessageSent event data.
+	expressions := []query.Expression{query.Confidence(primitives.Finalized)}
+	for _, data := range cctpData {
```

Comment:
That would definitely require perf tests using database like the ones we have for evm usdc reader 
https://github.com/smartcontractkit/chainlink/blob/develop/core/capabilities/ccip/ccip_integration_tests/usdcreader/usdcreader_test.go